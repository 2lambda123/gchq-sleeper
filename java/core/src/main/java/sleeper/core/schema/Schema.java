/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    public Schema() {
        this.rowKeyFields = new ArrayList<>();
        this.sortKeyFields = new ArrayList<>();
        this.valueFields = new ArrayList<>();
    }

    private Schema(Builder builder) {
        rowKeyFields = new ArrayList<>(builder.validRowKeyFields());
        sortKeyFields = new ArrayList<>(builder.validSortKeyFields());
        valueFields = new ArrayList<>(builder.validValueFields());
        validateNoDuplicates(Stream.concat(rowKeyFields.stream(),
                Stream.concat(sortKeyFields.stream(), valueFields.stream())));
    }

    public static Builder builder() {
        return new Builder();
    }

    public void setRowKeyFields(List<Field> rowKeyFields) {
        validateRowKeys(rowKeyFields);
        this.rowKeyFields.clear();
        this.rowKeyFields.addAll(rowKeyFields);
    }

    public void setRowKeyFields(Field... rowKeyFields) {
        setRowKeyFields(Arrays.asList(rowKeyFields));
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return rowKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public Field getSingleRowKeyField() {
        if (rowKeyFields.size() != 1) {
            throw new IllegalStateException("Cannot get single row key field, have " + rowKeyFields.size());
        }
        return rowKeyFields.get(0);
    }

    public void setSortKeyFields(List<Field> sortKeyFields) {
        validateSortKeys(sortKeyFields);
        this.sortKeyFields.clear();
        this.sortKeyFields.addAll(sortKeyFields);
    }

    public void setSortKeyFields(Field... sortKeyFields) {
        setSortKeyFields(Arrays.asList(sortKeyFields));
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return sortKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public void setValueFields(List<Field> valueFields) {
        this.valueFields.clear();
        this.valueFields.addAll(valueFields);
    }

    public void setValueFields(Field... valueFields) {
        setValueFields(Arrays.asList(valueFields));
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<String> getRowKeyFieldNames() {
        return rowKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getSortKeyFieldNames() {
        return sortKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getValueFieldNames() {
        return valueFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getAllFieldNames() {
        List<String> allFieldNames = new ArrayList<>();
        allFieldNames.addAll(getRowKeyFieldNames());
        allFieldNames.addAll(getSortKeyFieldNames());
        allFieldNames.addAll(getValueFieldNames());
        return allFieldNames;
    }

    public List<Field> getAllFields() {
        List<Field> allFields = new ArrayList<>();
        allFields.addAll(rowKeyFields);
        allFields.addAll(sortKeyFields);
        allFields.addAll(valueFields);
        return allFields;
    }

    public Optional<Field> getField(String fieldName) {
        return getAllFields().stream()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }

        private List<Field> validRowKeyFields() {
            return validateRowKeys(rowKeyFields);
        }

        private List<Field> validSortKeyFields() {
            if (sortKeyFields == null) {
                return Collections.emptyList();
            } else {
                return validateSortKeys(sortKeyFields);
            }
        }

        private List<Field> validValueFields() {
            if (valueFields == null) {
                return Collections.emptyList();
            } else {
                return valueFields;
            }
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return fields;
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return fields;
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
