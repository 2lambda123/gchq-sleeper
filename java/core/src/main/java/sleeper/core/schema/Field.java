/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.Type;

import java.util.Objects;

/**
 * Describes a field in the {@link Schema}. It contains a {@link String} name
 * and a {@link Type}.
 */
public class Field {
    private String name;
    private Type type;
    
    public Field() {
    }
    
    public Field(String name, Type type) {
        setName(name);
        setType(type);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        Objects.requireNonNull(name, "name must not be null");
        this.name = name;
    }

    public Type getType() {
        return type;
    }

    public void setType(Type type) {
        Objects.requireNonNull(type, "type must not be null");
        this.type = type;
    }

    @Override
    public String toString() {
        return "Field{" + "name=" + name + ", type=" + type + '}';
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 31 * hash + Objects.hashCode(this.name);
        hash = 31 * hash + Objects.hashCode(this.type);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Field other = (Field) obj;
        if (!Objects.equals(this.name, other.name)) {
            return false;
        }
        if (!Objects.equals(this.type, other.type)) {
            return false;
        }
        return true;
    }
}
