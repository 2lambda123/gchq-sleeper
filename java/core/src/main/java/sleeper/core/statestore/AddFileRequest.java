/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.core.statestore;

import java.util.function.Consumer;

public class AddFileRequest {
    private final String filename;
    private final String partitionId;
    private final FileDataStatistics statistics;

    private AddFileRequest(Builder builder) {
        filename = builder.filename;
        partitionId = builder.partitionId;
        statistics = builder.statistics;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static AddFileRequest addFile(Consumer<Builder> config) {
        Builder builder = builder();
        config.accept(builder);
        return builder.build();
    }

    public FileInfoV2 buildFileInfo() {
        return FileInfoV2.builder()
                .filename(filename)
                .partitionId(partitionId)
                .statistics(statistics)
                .build();
    }

    public static final class Builder {
        private String filename;
        private String partitionId;
        private FileDataStatistics statistics;

        private Builder() {
        }

        public Builder filename(String filename) {
            this.filename = filename;
            return this;
        }

        public Builder partitionId(String partitionId) {
            this.partitionId = partitionId;
            return this;
        }

        public Builder statistics(FileDataStatistics statistics) {
            this.statistics = statistics;
            return this;
        }

        public AddFileRequest build() {
            return new AddFileRequest(this);
        }
    }
}
