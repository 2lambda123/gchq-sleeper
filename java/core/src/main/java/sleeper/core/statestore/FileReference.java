/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore;

import java.time.Instant;
import java.util.Objects;

/**
 * Stores metadata about a reference to a physical file, such as its filename, which partition it is in,
 * the number of records in this section of the file, and optionally a job id indicating which compaction
 * job is responsible for compacting it.
 */
public class FileReference {

    private final String filename;
    private final String partitionId;
    private final Long numberOfRecords;
    private final String jobId;
    private final Instant lastStateStoreUpdateTime;
    private final boolean countApproximate;
    private final boolean onlyContainsDataForThisPartition;

    private FileReference(Builder builder) {
        filename = Objects.requireNonNull(builder.filename, "filename must not be null");
        partitionId = Objects.requireNonNull(builder.partitionId, "partitionId must not be null");
        numberOfRecords = Objects.requireNonNull(builder.numberOfRecords, "numberOfRecords must not be null");
        jobId = builder.jobId;
        lastStateStoreUpdateTime = builder.lastStateStoreUpdateTime;
        countApproximate = builder.countApproximate;
        onlyContainsDataForThisPartition = builder.onlyContainsDataForThisPartition;
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getFilename() {
        return filename;
    }

    public String getJobId() {
        return jobId;
    }

    public Instant getLastStateStoreUpdateTime() {
        return lastStateStoreUpdateTime;
    }

    public String getPartitionId() {
        return partitionId;
    }

    public Long getNumberOfRecords() {
        return numberOfRecords;
    }

    public boolean isCountApproximate() {
        return countApproximate;
    }

    public boolean onlyContainsDataForThisPartition() {
        return onlyContainsDataForThisPartition;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        FileReference fileReference = (FileReference) o;
        return countApproximate == fileReference.countApproximate && onlyContainsDataForThisPartition == fileReference.onlyContainsDataForThisPartition && Objects.equals(filename, fileReference.filename) && Objects.equals(partitionId, fileReference.partitionId) && Objects.equals(numberOfRecords, fileReference.numberOfRecords) && Objects.equals(jobId, fileReference.jobId) && Objects.equals(lastStateStoreUpdateTime, fileReference.lastStateStoreUpdateTime);
    }

    @Override
    public int hashCode() {
        return Objects.hash(filename, partitionId, numberOfRecords, jobId, lastStateStoreUpdateTime, countApproximate, onlyContainsDataForThisPartition);
    }

    @Override
    public String toString() {
        return "FileReference{" +
                "filename='" + filename + '\'' +
                ", partitionId='" + partitionId + '\'' +
                ", numberOfRecords=" + numberOfRecords +
                ", jobId='" + jobId + '\'' +
                ", lastStateStoreUpdateTime=" + lastStateStoreUpdateTime +
                ", countApproximate=" + countApproximate +
                ", onlyContainsDataForThisPartition=" + onlyContainsDataForThisPartition +
                '}';
    }

    public Builder toBuilder() {
        return new Builder()
                .filename(filename)
                .partitionId(partitionId)
                .numberOfRecords(numberOfRecords)
                .jobId(jobId)
                .lastStateStoreUpdateTime(lastStateStoreUpdateTime)
                .countApproximate(countApproximate)
                .onlyContainsDataForThisPartition(onlyContainsDataForThisPartition);
    }

    public static final class Builder {
        private String filename;
        private String partitionId;
        private Long numberOfRecords;
        private String jobId;
        private Instant lastStateStoreUpdateTime;
        private boolean countApproximate;
        private boolean onlyContainsDataForThisPartition;

        private Builder() {
        }

        public Builder filename(String filename) {
            this.filename = filename;
            return this;
        }

        public Builder partitionId(String partitionId) {
            this.partitionId = partitionId;
            return this;
        }

        public Builder numberOfRecords(Long numberOfRecords) {
            this.numberOfRecords = numberOfRecords;
            return this;
        }

        public Builder jobId(String jobId) {
            this.jobId = jobId;
            return this;
        }

        public Builder lastStateStoreUpdateTime(Instant lastStateStoreUpdateTime) {
            this.lastStateStoreUpdateTime = lastStateStoreUpdateTime;
            return this;
        }

        public Builder countApproximate(boolean countApproximate) {
            this.countApproximate = countApproximate;
            return this;
        }

        public Builder onlyContainsDataForThisPartition(boolean onlyContainsDataForThisPartition) {
            this.onlyContainsDataForThisPartition = onlyContainsDataForThisPartition;
            return this;
        }

        public FileReference build() {
            return new FileReference(this);
        }
    }
}
