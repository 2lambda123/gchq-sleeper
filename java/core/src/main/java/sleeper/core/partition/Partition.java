/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.range.Region;
import sleeper.core.range.RegionCanonicaliser;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.PrimitiveType;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * A Partition is a Region of key space, with additional information that allows
 * the position of the partition in the tree of partitions to be identified,
 * e.g. it has a unique id, the id of its parent partition, the id of any child
 * partitions, and if it has been split the dimension it was split on.
 * <p>
 * The {@link Region} must be in canonical form, i.e. all the ranges must contain
 * their minimum but not contain their maximum.
 */
public class Partition {
    private List<PrimitiveType> rowKeyTypes;
    private Region region;
    private String id;
    private boolean leafPartition;
    private String parentPartitionId;
    private List<String> childPartitionIds;
    private int dimension = -1; // -1 used to indicate that it has not been split yet; when it has been split, indicates which dimension was used to split on.

    public Partition() {
    }

    public Partition(
            List<PrimitiveType> rowKeyTypes,
            Region region,
            String id,
            boolean leafPartition,
            String parentPartitionId,
            List<String> childPartitionIds,
            int dimension) {
        setRowKeyTypes(rowKeyTypes);
        setRegion(region);
        setId(id);
        setLeafPartition(leafPartition);
        setParentPartitionId(parentPartitionId);
        setChildPartitionIds(childPartitionIds);
        setDimension(dimension);
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return rowKeyTypes;
    }

    public void setRowKeyTypes(List<PrimitiveType> rowKeyTypes) {
        this.rowKeyTypes = rowKeyTypes;
    }

    public void setRowKeyTypes(PrimitiveType... rowKeyTypes) {
        this.rowKeyTypes = new ArrayList<>();
        for (PrimitiveType type : rowKeyTypes) {
            this.rowKeyTypes.add(type);
        }
    }

    public Region getRegion() {
        return region;
    }

    public void setRegion(Region region) {
        if (!RegionCanonicaliser.isRegionInCanonicalForm(region)) {
            throw new IllegalArgumentException("Region must be in canonical form");
        }
        this.region = region;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public boolean isLeafPartition() {
        return leafPartition;
    }

    public void setLeafPartition(boolean leafPartition) {
        this.leafPartition = leafPartition;
    }

    public String getParentPartitionId() {
        return parentPartitionId;
    }

    public void setParentPartitionId(String parentPartitionId) {
        this.parentPartitionId = parentPartitionId;
    }

    public List<String> getChildPartitionIds() {
        if (null == childPartitionIds) {
            childPartitionIds = new ArrayList<>();
        }
        return childPartitionIds;
    }

    public void setChildPartitionIds(List<String> childPartitionIds) {
        this.childPartitionIds = new ArrayList<>();
        if (null != childPartitionIds) {
            this.childPartitionIds.addAll(childPartitionIds);
        }
    }

    public int getDimension() {
        return dimension;
    }

    public void setDimension(int dimension) {
        this.dimension = dimension;
    }

    public boolean isRowKeyInPartition(Schema schema, Key rowKey) {
        return region.isKeyInRegion(schema, rowKey);
    }

    public boolean doesRegionOverlapPartition(Region otherRegion) {
        return region.doesRegionOverlap(otherRegion);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Partition partition = (Partition) o;

        return Objects.equals(rowKeyTypes, partition.rowKeyTypes)
                && leafPartition == partition.leafPartition
                && Objects.equals(region, partition.region)
                && Objects.equals(id, partition.id)
                && Objects.equals(parentPartitionId, partition.parentPartitionId)
                && Objects.equals(getChildPartitionIds(), partition.getChildPartitionIds())
                && dimension == partition.getDimension();
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyTypes, region, id, leafPartition,
                parentPartitionId, getChildPartitionIds(), dimension);
    }

    @Override
    public String toString() {
        return "Partition{"
                + "rowKeyTypes=" + rowKeyTypes
                + ", region=" + region
                + ", id='" + id + '\''
                + ", leafPartition=" + leafPartition
                + ", parentPartitionId='" + parentPartitionId + '\''
                + ", childPartitionIds=" + getChildPartitionIds()
                + ", dimension=" + dimension
                + '}';
    }
}
