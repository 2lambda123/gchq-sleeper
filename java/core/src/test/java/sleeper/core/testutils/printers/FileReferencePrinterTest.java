/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.core.testutils.printers;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.StringType;
import sleeper.core.statestore.FileReference;
import sleeper.core.statestore.FileReferenceFactory;
import sleeper.core.table.TableIdentity;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
import static sleeper.core.statestore.FilesReportTestHelper.activeAndReadyForGCFiles;
import static sleeper.core.statestore.FilesReportTestHelper.activeFiles;
import static sleeper.core.statestore.FilesReportTestHelper.noFiles;
import static sleeper.core.statestore.SplitFileReference.referenceForChildPartition;

public class FileReferencePrinterTest {

    private final Schema schema = schemaWithKey("key", new StringType());
    private final PartitionsBuilder partitions = new PartitionsBuilder(schema);

    @Nested
    @DisplayName("Active files")
    class ActiveFiles {

        @Test
        void shouldPrintMultipleFilesInPartition() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50");

            // When
            FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), List.of(
                    fileReferenceFactory.partitionFile("L", 10),
                    fileReferenceFactory.partitionFile("L", 20),
                    fileReferenceFactory.partitionFile("R", 30),
                    fileReferenceFactory.partitionFile("R", 40)));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintPartialFiles() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50");
            FileReference file1 = fileReferenceFactory().rootFile("a.parquet", 100);
            FileReference file2 = fileReferenceFactory().rootFile("a.parquet", 200);

            // When
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), List.of(
                    referenceForChildPartition(file1, "L"),
                    referenceForChildPartition(file2, "L"),
                    referenceForChildPartition(file1, "R"),
                    referenceForChildPartition(file2, "R")));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintFilesOnLeaves() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50")
                    .splitToNewChildren("L", "LL", "LR", "row-25")
                    .splitToNewChildren("R", "RL", "RR", "row-75")
                    .splitToNewChildren("LL", "LLL", "LLR", "row-12")
                    .splitToNewChildren("LR", "LRL", "LRR", "row-37")
                    .splitToNewChildren("RL", "RLL", "RLR", "row-62")
                    .splitToNewChildren("RR", "RRL", "RRR", "row-87");

            // When
            FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), List.of(
                    fileReferenceFactory.partitionFile("LLL", 12),
                    fileReferenceFactory.partitionFile("LLR", 13),
                    fileReferenceFactory.partitionFile("LRL", 12),
                    fileReferenceFactory.partitionFile("LRR", 13),
                    fileReferenceFactory.partitionFile("RLL", 12),
                    fileReferenceFactory.partitionFile("RLR", 13),
                    fileReferenceFactory.partitionFile("RRL", 12),
                    fileReferenceFactory.partitionFile("RRR", 13)));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldOrderFilesByPartitionLocationInTree() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50")
                    .splitToNewChildren("L", "LL", "LR", "row-25")
                    .splitToNewChildren("R", "RL", "RR", "row-75")
                    .splitToNewChildren("LL", "LLL", "LLR", "row-12")
                    .splitToNewChildren("LR", "LRL", "LRR", "row-37")
                    .splitToNewChildren("RL", "RLL", "RLR", "row-62")
                    .splitToNewChildren("RR", "RRL", "RRR", "row-87");

            // When
            FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), List.of(
                    fileReferenceFactory.partitionFile("L", 50),
                    fileReferenceFactory.partitionFile("LRL", 12),
                    fileReferenceFactory.partitionFile("root", 100),
                    fileReferenceFactory.partitionFile("RLL", 12),
                    fileReferenceFactory.partitionFile("RR", 12),
                    fileReferenceFactory.partitionFile("LLL", 13),
                    fileReferenceFactory.partitionFile("R", 12),
                    fileReferenceFactory.partitionFile("RRL", 13),
                    fileReferenceFactory.partitionFile("LLR", 25),
                    fileReferenceFactory.partitionFile("RLR", 50),
                    fileReferenceFactory.partitionFile("RRR", 100)));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldRenamePartitionsByLocation() {
            // Given
            partitions.rootFirst("base")
                    .splitToNewChildren("base", "l", "r", "row-50")
                    .splitToNewChildren("l", "ll", "lr", "row-25")
                    .splitToNewChildren("r", "rl", "rr", "row-75")
                    .splitToNewChildren("ll", "1", "2", "row-12")
                    .splitToNewChildren("lr", "3", "4", "row-37")
                    .splitToNewChildren("rl", "5", "6", "row-62")
                    .splitToNewChildren("rr", "7", "8", "row-87");

            // When
            FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), List.of(
                    fileReferenceFactory.partitionFile("1", 12),
                    fileReferenceFactory.partitionFile("2", 13),
                    fileReferenceFactory.partitionFile("3", 12),
                    fileReferenceFactory.partitionFile("4", 13),
                    fileReferenceFactory.partitionFile("5", 12),
                    fileReferenceFactory.partitionFile("6", 13),
                    fileReferenceFactory.partitionFile("7", 12),
                    fileReferenceFactory.partitionFile("8", 13),
                    fileReferenceFactory.partitionFile("ll", 25),
                    fileReferenceFactory.partitionFile("l", 50),
                    fileReferenceFactory.partitionFile("base", 100)));

            // Then see approved output
            Approvals.verify(printed);
        }
    }

    @Nested
    @DisplayName("Multiple tables")
    class MultipleTables {

        @Test
        void shouldPrintFilesOnceWhenTwoTablesAreIdentical() {
            // Given
            partitions.rootFirst("root");
            List<FileReference> files = List.of(fileReferenceFactory().partitionFile("root", 10));

            // When
            String printed = FileReferencePrinter.printTableFilesExpectingIdentical(
                    Map.of("table-1", partitions.buildTree(), "table-2", partitions.buildTree()),
                    Map.of("table-1", files, "table-2", files));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintDifferentFilesForOneTable() {
            // Given
            partitions.rootFirst("root");
            List<FileReference> files1 = List.of(fileReferenceFactory().partitionFile("root", 10));
            List<FileReference> files2 = List.of(fileReferenceFactory().partitionFile("root", 20));

            // When
            String printed = FileReferencePrinter.printTableFilesExpectingIdentical(
                    Map.of("table-1", partitions.buildTree(), "table-2", partitions.buildTree(), "table-3", partitions.buildTree()),
                    Map.of("table-1", files1, "table-2", files2, "table-3", files1));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintOnlyOneTable() {
            // Given
            partitions.rootFirst("root");
            List<FileReference> files = List.of(fileReferenceFactory().partitionFile("root", 10));

            // When
            String printed = FileReferencePrinter.printTableFilesExpectingIdentical(
                    Map.of("table-1", partitions.buildTree()),
                    Map.of("table-1", files));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintExpectedForTables() {
            // Given
            partitions.rootFirst("root");
            List<FileReference> files = List.of(fileReferenceFactory().partitionFile("root", 10));

            // When
            String printed = FileReferencePrinter.printExpectedFilesForAllTables(
                    List.of(table("table-1"), table("table-2")), partitions.buildTree(), files);

            // Then
            assertThat(printed).isEqualTo(FileReferencePrinter.printTableFilesExpectingIdentical(
                    Map.of("table-1", partitions.buildTree(), "table-2", partitions.buildTree()),
                    Map.of("table-1", files, "table-2", files)));
        }
    }

    @Nested
    @DisplayName("All files including unreferenced")
    class AllFiles {

        @Test
        void shouldPrintReferencedAndUnreferencedFiles() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50");

            // When
            FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(),
                    activeAndReadyForGCFiles(
                            List.of(fileReferenceFactory.partitionFile("L", 10)),
                            List.of("oldFile1.parquet", "oldFile2.parquet")));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintFileSplitAcrossPartitions() {
            // Given
            partitions.rootFirst("root")
                    .splitToNewChildren("root", "L", "R", "row-50");

            // When
            FileReference rootFile = fileReferenceFactory().rootFile(20);
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(),
                    activeFiles(
                            referenceForChildPartition(rootFile, "L"),
                            referenceForChildPartition(rootFile, "R")));

            // Then see approved output
            Approvals.verify(printed);
        }

        @Test
        void shouldPrintNoFiles() {
            // Given
            partitions.rootFirst("root");

            // When
            String printed = FileReferencePrinter.printFiles(partitions.buildTree(), noFiles());

            // Then see approved output
            Approvals.verify(printed);
        }
    }

    private TableIdentity table(String name) {
        return TableIdentity.uniqueIdAndName(name, name);
    }

    private FileReferenceFactory fileReferenceFactory() {
        return FileReferenceFactory.from(partitions.buildTree());
    }
}
