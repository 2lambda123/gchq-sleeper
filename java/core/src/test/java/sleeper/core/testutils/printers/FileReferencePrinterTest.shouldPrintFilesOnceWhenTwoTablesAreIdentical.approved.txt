Same for 2 tables
Active files:
Partition at root: 10 records in file

